### DISCUSSION TITLE
WiringOp missing a library?

### DISCUSSION THREAD

**Discussion by life8853** [Joshua-Riek/ubuntu-rockchip#181] [Q&A] [answered]
So i installed the latest version (1.16) and tried to mess with some gpio pins to plug my fan into. 

Command _gpio readall_ works but when compiling a C script `fan-control.c:4:10: fatal error: wiringPi.h: No such file or directory` error occures.

Also if someone has a good explanation how to use gpio and how am i supposed to plug a 4 pin fan when there is no 4 fan pinout on the board?

**Comment by Joshua-Riek** [comment, marked as answer]
Make sure the package `libwiringpi-opi-dev` is installed, I double-checked the package and I see the file `wiringPi.h` under the path `/usr/include`. Maybe you need to set /usr/include to your header search path.

**Reply by Joshua-Riek** [reply]
By the way if you want to look at the package / source, its available on my ppa: https://launchpad.net/~jjriek/+archive/ubuntu/rockchip/+packages

**Comment by life8853** [comment]
Oh, so you have to install it yourself, i thought it was one package with gpio command, so i deducted that the library should have also been there. Thanks!

**Reply by Joshua-Riek** [reply]
If this library is not included by default, ill make sure that it is included in future releases.

**Reply by life8853** [reply]
@Joshua-Riek  it is not installed by default, although installing it caused a diffrerent problem, now the gcc can't link it

```
    ubuntu@OrangePi:~/opi5-fan-control$ gcc -o fan-control fan-control.c -lwiringpi
    /usr/bin/ld: cannot find -lwiringpi: No such file or directory
    collect2: error: ld returned 1 exit status
```

Also i don't know why Github doesn't make the code block 3 lines?

**Reply by Joshua-Riek** [reply]
Strange, can you upload your fan-control.c file? Ill try a few things on my end. 
For code blocks on GitHub take a look here: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code

**Reply by life8853** [reply]
I was trying to compile this: https://github.com/bastilmr/opi5-fan-control/blob/master/fan-control.c

**Comment by Joshua-Riek** [comment]
I found the issue; the packaging has two broken symlinks!
```
ubuntu@ubuntu-desktop:~/work$ ls /usr/lib/libwiringPi* -la
lrwxrwxrwx 1 root root     91 May  1 21:01 /usr/lib/libwiringPiDev.so -> /build/wiringpi-opi-XbjFCu/wiringpi-opi-2.275~iwj/debian/tmp/usr/lib/libwiringPiDev.so.2.48
lrwxrwxrwx 1 root root     22 May  1 21:01 /usr/lib/libwiringPiDev.so.2 -> libwiringPiDev.so.2.48
-rw-r--r-- 1 root root  29328 May  1 21:01 /usr/lib/libwiringPiDev.so.2.48
lrwxrwxrwx 1 root root     88 May  1 21:01 /usr/lib/libwiringPi.so -> /build/wiringpi-opi-XbjFCu/wiringpi-opi-2.275~iwj/debian/tmp/usr/lib/libwiringPi.so.2.48
lrwxrwxrwx 1 root root     19 May  1 21:01 /usr/lib/libwiringPi.so.2 -> libwiringPi.so.2.48
-rw-r--r-- 1 root root 123680 May  1 21:01 /usr/lib/libwiringPi.so.2.48
```

Here is a quick fix for the broken symlinks.
```
ubuntu@ubuntu-desktop:~/work$ cd /usr/lib
ubuntu@ubuntu-desktop:/usr/lib$ sudo rm /usr/lib/libwiringPiDev.so 
ubuntu@ubuntu-desktop:/usr/lib$ sudo ln -s libwiringPiDev.so.2.48  libwiringPiDev.so 
ubuntu@ubuntu-desktop:/usr/lib$ sudo rm libwiringPi.so
ubuntu@ubuntu-desktop:/usr/lib$ sudo ln -s libwiringPi.so.2.48 libwiringPi.so
```

And now the fan control program will compile just fine.
```
ubuntu@ubuntu-desktop:/usr/lib$ cd ~/work/
ubuntu@ubuntu-desktop:~/work$ gcc -o main main.c -lwiringPi
```

**Reply by life8853** [reply]
Oh, thank you very much, is this something that you can fix in the next release or do i have to note it down somewhere?

**Reply by Joshua-Riek** [reply]
I can fix it with a simple `sudo apt update && sudo apt upgrade` on your end. I need to debug the packaging; I have no clue why that symlink is broken. 

If you would like, you can fix it on your opi5 with the below commands:
```
cd /usr/lib
sudo rm /usr/lib/libwiringPiDev.so 
sudo ln -s libwiringPiDev.so.2.48  libwiringPiDev.so 
sudo rm libwiringPi.so
sudo ln -s libwiringPi.so.2.48 libwiringPi.so
```

**Reply by Joshua-Riek** [reply]
I fixed the packaging, and the build on my ppa has finished. It should be available to download with an apt update in 1-2 hours.

https://launchpad.net/~jjriek/+archive/ubuntu/rockchip/+build/26277590
